File: frontend/src/types/market.ts
export type MarketBrief = {
  summary: string;
  why_this_matters_now: string;
  what_changed: string;
  bull_case: string;
  base_case: string;
  bear_case: string;
  catalysts: string;
  drivers: string;
  risks: string;
  what_would_change_probability: string;
  sources_json: Array<{ label: string; type: string }>;
  recent_headlines_json: Array<{
    title: string;
    source: string;
    published_at: string;
    why_it_matters: string;
  }>;
  generated_at: string;
};

export type MarketSnapshot = {
  id: number;
  captured_at: string;
  price: number;
  volume: number | null;
  open_interest: number | null;
  metadata_json: Record<string, unknown> | null;
};

export type MarketTimelineEntry = {
  id: number;
  occurred_at: string;
  move: number | null;
  price_after_move: number | null;
  reason: string;
  linked_label: string | null;
  linked_type: string | null;
  linked_url: string | null;
  metadata_json: Record<string, unknown> | null;
};

export type Market = {
  id: number;
  external_id: string | null;
  source: string;
  title: string;
  slug: string;
  category: string;
  status: string;
  close_time: string | null;
  last_price: number;
  probability_change: number;
  volume: number | null;
  open_interest: number | null;
  description: string | null;
  metadata_json: Record<string, unknown> | null;
  brief: MarketBrief | null;
};

export type MarketDetail = Market & {
  snapshots: MarketSnapshot[];
  timeline_entries: MarketTimelineEntry[];
};

File: backend/app/schemas/market.py
import json
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_serializer


class MarketSnapshotRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    captured_at: datetime
    price: float
    volume: int | None = None
    open_interest: int | None = None
    metadata_json: str | None = None

    @field_serializer("metadata_json")
    def serialize_metadata(self, value: str | None) -> Any:
        if not value:
            return None
        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return value


class MarketTimelineEntryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    occurred_at: datetime
    move: float | None = None
    price_after_move: float | None = None
    reason: str
    linked_label: str | None = None
    linked_type: str | None = None
    linked_url: str | None = None
    metadata_json: str | None = None

    @field_serializer("metadata_json")
    def serialize_metadata(self, value: str | None) -> Any:
        if not value:
            return None
        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return value


class MarketBriefRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    summary: str
    why_this_matters_now: str
    what_changed: str
    bull_case: str
    base_case: str
    bear_case: str
    catalysts: str
    drivers: str
    risks: str
    what_would_change_probability: str
    sources_json: str | None = None
    recent_headlines_json: str | None = None
    generated_at: datetime

    @field_serializer("sources_json")
    def serialize_sources(self, value: str | None) -> Any:
        if not value:
            return []
        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return []

    @field_serializer("recent_headlines_json")
    def serialize_recent_headlines(self, value: str | None) -> Any:
        if not value:
            return []
        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return []


class MarketListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    external_id: str | None = None
    source: str
    title: str
    slug: str
    category: str
    status: str
    close_time: datetime | None = None
    last_price: float
    probability_change: float
    volume: int | None = None
    open_interest: int | None = None
    description: str | None = None
    metadata_json: str | None = None
    brief: MarketBriefRead | None = None

    @field_serializer("metadata_json")
    def serialize_metadata(self, value: str | None) -> Any:
        if not value:
            return None
        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return value


class MarketDetail(MarketListItem):
    snapshots: list[MarketSnapshotRead] = Field(default_factory=list)
    timeline_entries: list[MarketTimelineEntryRead] = Field(default_factory=list)
